app/code/Magento/Store/Controller/Store/SwitchAction.php

```diff
@@ -82,12 +82,12 @@ public function execute()
             $error = __('Requested store is inactive');
         } catch (NoSuchEntityException $e) {
             $error = __('Requested store is not found');
-        }
-
-        if (isset($error)) {
-            $this->messageManager->addError($error);
-            $this->getResponse()->setRedirect($this->_redirect->getRedirectUrl());
-            return;
+        } finally {
+            if (isset($error)) {
+                $this->messageManager->addError($error);
+                $this->getResponse()->setRedirect($this->_redirect->getRedirectUrl());
+                return;
```

Comment:
not supported by HHVM https://3v4l.org/LZ0Mq
that's why it's implemented as a flag in Magento codebase